File: Leetcode/61-Rotate-List/RotateList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if (!head || !head.next) {
        return head
    }
    var getLength = function(head) {
        var cur = head
        var len = 0
        while (cur) {
            cur = cur.next
            len++
        }
        return len
    }
    var len = getLength(head);
    k = k%len
    if (k==0) {
        return head
    }
    var dummy = new ListNode(0)
    dummy.next = head
    fast = dummy
    while (k>0) {
        fast = fast.next
        k--;
    }
    var slow = dummy
    while (fast.next) {
        slow = slow.next
        fast = fast.next
    }
    dummy.next = slow.next
    slow.next = null
    fast.next = head
    return dummy.next
};

File: Leetcode/61-Rotate-List/RotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        int len = getLength(head);
        k = k%len;
        if (k==0) {
            return head;
        }
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ListNode* fast = dummy;
        while (k>0) {
            fast = fast->next;
            k--;
        }
        ListNode* slow = dummy;
        while (fast->next!=NULL) {
            slow=slow->next;
            fast=fast->next;
        }
        dummy->next = slow->next;
        slow->next = NULL;
        fast->next = head;
        return dummy->next;
    }
private:
    int getLength(ListNode* head) {
        ListNode* cur = head;
        int len = 0;
        while (cur!=NULL) {
            cur = cur->next;
            len++;
        }
        return len;
    }
};
